src/neo/Network/P2P/Payloads/StateRoot.cs

```diff
@@ -0,0 +1,132 @@
+using Neo.Cryptography;
+using Neo.Cryptography.ECC;
+using Neo.IO;
+using Neo.IO.Json;
+using Neo.Persistence;
+using Neo.SmartContract;
+using Neo.SmartContract.Native;
+using Neo.SmartContract.Native.Designate;
+using System;
+using System.IO;
+
+namespace Neo.Network.P2P.Payloads
+{
+    public class StateRoot : ICloneable<StateRoot>, IVerifiable, ISerializable
+    {
+        public byte Version;
+        public uint Index;
+        public UInt256 RootHash;
+        public Witness Witness;
+
+        private UInt256 _hash = null;
+
+        public UInt256 Hash
+        {
+            get
+            {
+                if (_hash == null)
+                {
+                    _hash = new UInt256(Crypto.Hash256(this.GetHashData()));
+                }
+                return _hash;
+            }
+        }
+
+        Witness[] IVerifiable.Witnesses
+        {
+            get
+            {
+                return new[] { Witness };
+            }
+            set
+            {
+                if (value.Length != 1) throw new ArgumentException();
+                Witness = value[0];
+            }
+        }
+
+        public int Size =>
+            sizeof(byte) +      //Version
+            sizeof(uint) +      //Index
+            UInt256.Length +    //RootHash
+            1 +                 //Witness array count
+            Witness.Size;       //Witness
+
+        public StateRoot Clone()
+        {
+            return new StateRoot
+            {
+                Version = Version,
+                Index = Index,
+                RootHash = RootHash,
+                Witness = Witness,
+            };
+        }
+
+        public void FromReplica(StateRoot replica)
+        {
+            Version = replica.Version;
+            Index = replica.Index;
+            RootHash = replica.RootHash;
+            Witness = replica.Witness;
+        }
+
+        public void Deserialize(BinaryReader reader)
+        {
+            this.DeserializeUnsigned(reader);
+            Witness[] arr = reader.ReadSerializableArray<Witness>();
+            if (arr.Length < 1)
+                Witness = null;
```

Comment:
Does we need states without witness?

Reply:
Yes. The state root calculated locally has no witness.

Reply:
Maybe we should have two payloads, signed and unsigned, in order to prevent send of unsigned state roots.

Reply:
I add witness null check. But how can extra payload prevent sending unsigned roots?

Reply:
We will send and receive only signed roots, the unsigned payload will be used locally, without p2p message.

Reply:
I want to store `StateRoot` with `null` witness first. After validated `StateRoot` received, I can replace the `null` witness with signed witness.